operator/rpc_client.go

```diff
@@ -2,421 +2,259 @@ package operator
+		logger.Fatal("Registry coordinator address from aggregator does not match the one in the config", "aggregator", aggregatorRegistryCoordinatorAddress, "config", expectedRegistryCoordinatorAddress.String())
+		return nil, errors.New("mismatching registry coordinator address from aggregator")
 	}
 
-	c.rpcClient = client
-
-	return nil
+	return client, nil
 }
 
-func (c *AggregatorRpcClient) InitializeClientIfNotExist() error {
-	c.rpcClientLock.RLock()
-	if c.rpcClient != nil {
-		c.rpcClientLock.RUnlock()
-		return nil
-	}
-	c.rpcClientLock.RUnlock()
+type RetryStrategy func(submittedAt time.Time, err error) bool
 
-	return c.dialAggregatorRpcClient()
+func NeverRetry(_ time.Time, _ error) bool {
+	return false
 }
 
-func isShutdownOrNetworkError(err error) bool {
-	if err == rpc.ErrShutdown {
-		return true
-	}
+func AlwaysRetry(_ time.Time, _ error) bool {
+	return true
+}
 
-	if _, ok := err.(*net.OpError); ok {
+func RetryWithDelay(delay time.Duration) RetryStrategy {
+	return func(submittedAt time.Time, err error) bool {
+		time.Sleep(delay)
 		return true
 	}
-
-	return false
 }
 
-func (c *AggregatorRpcClient) handleRpcError(err error) {
-	if isShutdownOrNetworkError(err) {
-		go c.handleRpcShutdown()
+func RetryIfRecentEnough(ttl time.Duration) RetryStrategy {
+	return func(submittedAt time.Time, err error) bool {
+		return time.Since(submittedAt) < ttl
 	}
 }
 
-func (c *AggregatorRpcClient) handleRpcShutdown() {
-	c.rpcClientLock.Lock()
-	defer c.rpcClientLock.Unlock()
-
-	if c.rpcClient != nil {
-		c.logger.Info("Closing RPC client due to shutdown")
-
-		err := c.rpcClient.Close()
-		if err != nil {
-			c.logger.Error("Error closing RPC client", "err", err)
-		}
-
-		c.rpcClient = nil
+func RetryAtMost(retries int) RetryStrategy {
+	retryCount := 0
+	return func(_ time.Time, err error) bool {
+		result := retryCount < retries
+		retryCount++
+		return result
 	}
 }
 
-func (c *AggregatorRpcClient) onTick() {
-	for {
-		<-c.resendTicker.C
-
-		err := c.InitializeClientIfNotExist()
-		if err != nil {
-			c.logger.Error("Error initializing client", "err", err)
-			continue
-		}
-
-		c.unsentMessagesLock.Lock()
-		if len(c.unsentMessages) == 0 {
-			c.unsentMessagesLock.Unlock()
-			continue
-		}
-		c.unsentMessagesLock.Unlock()
-
-		c.tryResendFromDeque()
+func RetryAnd(s1 RetryStrategy, s2 RetryStrategy) RetryStrategy {
+	return func(submittedAt time.Time, err error) bool {
+		return s1(submittedAt, err) && s2(submittedAt, err)
 	}
 }
 
-// Expected to be called with initialized client.
-func (c *AggregatorRpcClient) tryResendFromDeque() {
-	c.rpcClientLock.RLock()
-	defer c.rpcClientLock.RUnlock()
+// By defaul, retry with a delay of 2 seconds between calls,
+// at most 10 times, and only if the error is recent enough (24 hours)
+// TODO: Discuss the "recent enough" part
+var DefaultAggregatorRpcRetryStrategy RetryStrategy = RetryAnd(
+	RetryWithDelay(2*time.Second), RetryAnd(
+		RetryAtMost(10),
+		RetryIfRecentEnough(24*time.Hour)))
```

Comment:
What would be a reasonable TTL for messages? This is a very conservative default just to start the discussion.